examples/DataApi-Voting-poll/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -0,0 +1,360 @@
+      ().validNec
+    } else {
+      InvalidAddress.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+
+  def validateIfUserAlreadyVoted(maybeState: Option[PollState], address: Address): DataApplicationValidationErrorOr[Unit] = {
+    maybeState match {
+      case Some(value) =>
+        val pollVotes = value.votes
+        pollVotes.get(address) match {
+          case Some(_) => RepeatedVote.asInstanceOf[DataApplicationValidationError].invalidNec
+          case None => ().validNec
+        }
+      case None => ().validNec
+    }
+  }
+
+  def validateSnapshotCreatePoll(snapshotOrdinal: SnapshotOrdinal, update: CreatePollUpdate): DataApplicationValidationErrorOr[Unit] = {
+    if (update.endSnapshotOrdinal < snapshotOrdinal.value.value) {
+      InvalidEndSnapshot.asInstanceOf[DataApplicationValidationError].invalidNec
+    } else {
+      ().validNec
+    }
+  }
+
+  def validatePollSnapshotInterval(lastSnapshotOrdinal: SnapshotOrdinal, currentState: State, voteInPoll: VoteInPollUpdate): DataApplicationValidationErrorOr[Unit] = {
+    val poll = currentState.polls.get(voteInPoll.pollId)
+    poll match {
+      case Some(value) =>
+        if (value.endSnapshotOrdinal < lastSnapshotOrdinal.value.value) {
+          ClosedPool.asInstanceOf[DataApplicationValidationError].invalidNec
+        } else {
+          ().validNec
+        }
+      case None => PollDoesNotExists.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+}
+
+object Validations {
+  def createPollValidations(update: CreatePollUpdate, state: State, lastSnapshotOrdinal: Option[SnapshotOrdinal]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateCreatePollSnapshot = IO {
+      lastSnapshotOrdinal match {
+        case Some(value) => validateSnapshotCreatePoll(value, update)
+        case None => ().validNec
+      }
+    }
+
+    val validatePoll = IO {
+      val voteId = Hash.fromBytes(customUpdateSerialization(update))
+      validateIfPollAlreadyExists(state.polls.get(voteId.toString))
+    }
+
+    for {
+      validatedCreatePollSnapshot <- validateCreatePollSnapshot
+      validatedPoll <- validatePoll
+    } yield validatedCreatePollSnapshot.productR(validatedPoll)
+
+  }
+
+  def voteInPollValidations(update: VoteInPollUpdate, state: State, lastSnapshotOrdinal: Option[SnapshotOrdinal]): IO[DataApplicationValidationErrorOr[Unit]] = {
+    val validateSnapshotInterval = IO {
+      lastSnapshotOrdinal match {
+        case Some(value) => validatePollSnapshotInterval(value, state, update)
+        case None => ().validNec
+      }
+    }
+
+    val validatePoll = IO {
+      validateIfVotePollExists(state.polls.get(update.pollId))
+    }
+
+    val validateOption = IO {
+      validateIfOptionExists(state.polls.get(update.pollId), update.option)
+    }
+
+    val validateRepeatedVote = IO {
+      validateIfUserAlreadyVoted(state.polls.get(update.pollId), update.address)
+    }
+
+    for {
```

Comment:
As a note, it appears most of these underlying functions don't require IO, hence we're constructing an IO here and immediately doing a for comprehension flatmap syntax to remove them, making this somewhat redundant versus just using productR directly on the final immediate types.

The only reason to bind something into an IO should be for interaction with another IO already present, which most of these functions do not appear to have.